ingestion-beam/src/main/java/com/mozilla/telemetry/transforms/DecompressPayload.java

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package com.mozilla.telemetry.transforms;
+
+import com.google.common.primitives.Ints;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.zip.GZIPInputStream;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.commons.io.IOUtils;
+
+public class DecompressPayload
+    extends PTransform<PCollection<PubsubMessage>, PCollection<PubsubMessage>> {
+
+  /**
+   * Return a {@link PTransform} that performs gzip decompression on gzipped message payloads and
+   * passes other messages through unchanged.
+   *
+   * @param enabled if false, this transform will pass all messages through unchanged
+   */
+  public static DecompressPayload enabled(ValueProvider<Boolean> enabled) {
+    return new DecompressPayload(enabled);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> input) {
+    return input.apply(MapElements.via(fn));
+  }
+
+  ////////
+
+  private final ValueProvider<Boolean> enabled;
+
+  private DecompressPayload(ValueProvider<Boolean> enabled) {
+    this.enabled = enabled;
+  }
+
+  private class Fn extends SimpleFunction<PubsubMessage, PubsubMessage> {
+
+    private static final byte ZERO_BYTE = 0x00;
+
+    private final Counter compressedInput = Metrics.counter(Fn.class, "compressed-input");
+    private final Counter uncompressedInput = Metrics.counter(Fn.class, "uncompressed-input");
+
+    @Override
+    public PubsubMessage apply(PubsubMessage value) {
+      byte[] payload = value.getPayload();
+      if (enabled.isAccessible() && !enabled.get()) {
+        // Compression has been explicitly turned off in options, so return unchanged message.
+        return value;
+      } else if (!isGzip(payload)) {
```

Comment:
is detecting gzip here more efficient than just catching an IOException in the next block?

Reply:
I don't have evidence at this point to indicate how expensive it is to let decompression throw an exception on uncompressed input. I added this primarily due to observing metrics on the running Decoder job. This decompression step was the second most expensive step behind ParsePayload.

I'm open to reverting this if you're concerned about adding the complexity.

Reply:
Removed. We can measure this and add the implementation back as part of the zstd project if we take that up.